### ISSUE TITLE
HashSpi allocates single byte arrays on the single byte update path

### ISSUE THREAD

**Issue by jolynch** [corretto/amazon-corretto-crypto-provider#52]
While benchmarking ACCP on code that is more or less:
```
MessageDigest digest = MessageDigest.getInstance("MD5");
for (<many iterations>)
{
...
    digest.update(<single byte>);
}
```

I noticed that ACCP was generating significantly more garbage and was a lot slower than expected. I believe this is because `TemplateHashSpi` allocates a single byte array on the hot path of single byte updates:
https://github.com/corretto/amazon-corretto-crypto-provider/blob/b204b018f6aa5d42b4fee0d0a94a93994bede081/template-src/com/amazon/corretto/crypto/provider/TemplateHashSpi.java#L119-L121

Since the Spi contract is inherently not threadsafe, performance on use cases such as above could be improved signficantly by caching a single byte buffer like the TemplateHmacSpi does:
https://github.com/corretto/amazon-corretto-crypto-provider/blob/b204b018f6aa5d42b4fee0d0a94a93994bede081/template-src/com/amazon/corretto/crypto/provider/TemplateHmacSpi.java#L299-L304

An easy reproduction is to run the following with and without ACCP:
```java
import java.security.MessageDigest;

public final class Test
{
    public static void updateWithInt(MessageDigest digest, int val)
    {
        digest.update((byte) ((val >>> 24) & 0xFF));
        digest.update((byte) ((val >>> 16) & 0xFF));
        digest.update((byte) ((val >>>  8) & 0xFF));
        digest.update((byte) ((val >>> 0) & 0xFF));
    }

    public static void main(String[] args) throws Exception
    {
        int numRounds = 100000000;
        if (args.length  > 0) {
            numRounds = Integer.parseInt(args[0]);
        }
        System.out.println("Burn test of MD5");
        MessageDigest digest = MessageDigest.getInstance("MD5");
        System.out.println("Using Digest: " + digest.toString());
        long start = System.currentTimeMillis();
        for (int i = 0; i < numRounds ; i ++) {
            updateWithInt(digest, i);
        }
        long end = System.currentTimeMillis();

        System.out.println("Result:   " + digest.digest());
        System.out.println("Time(ms): " + (end - start));
    }
}
```

```
time java -Djava.security.properties=/path/to/amazon-corretto-crypto-provider.security -cp AmazonCorrettoCryptoProvider-1.1.0-linux-x86_64.jar:. Test 100000000
Burn test of MD5
Using Digest: MD5 Message Digest from AmazonCorrettoCryptoProvider, <initialized>

Result:   [B@6166e06f
Time(ms): 10358
java  -cp AmazonCorrettoCryptoProvider-1.1.0-linux-x86_64.jar:. Test 10000000  12.91s user 0.25s system 117% cpu 11.227 total
```
vs
```
time java -cp . Test 100000000 
Burn test of MD5
Using Digest: MD5 Message Digest from SUN, <initialized>

Result:   [B@2a139a55
Time(ms): 3878
java -cp . Test 100000000  3.99s user 0.02s system 101% cpu 3.945 total
```
Also using sjk we can see that the Corretto version is allocating close to 900 MiBps:
```
$ sjk ttop -o ALLOC -p $(pgrep -f Test)
2019-08-03T23:38:31.019-0700 Process summary 
  process cpu=107.39%
  application cpu=101.04% (user=99.38% sys=1.65%)
  other: cpu=6.35% 
  thread count: 12
  GC time=0.23% (young=0.23%, old=0.00%)
  heap allocation rate 842mb/s
  safe point rate: 1.6 (events/s) avg. safe point pause: 1.65ms
  safe point sync time: 0.01% processing time: 0.25% (wallclock time)
[000001] user=98.21% sys= 1.50% alloc=  842mb/s - main
[000016] user= 1.17% sys= 0.01% alloc=  324kb/s - RMI TCP Connection(1)-127.0.0.1
[000018] user= 0.00% sys= 0.13% alloc=  4461b/s - JMX server connection timeout 18
[000002] user= 0.00% sys= 0.00% alloc=     0b/s - Reference Handler
[000003] user= 0.00% sys= 0.00% alloc=     0b/s - Finalizer
[000004] user= 0.00% sys= 0.00% alloc=     0b/s - Signal Dispatcher
[000011] user= 0.00% sys= 0.00% alloc=     0b/s - ForkJoinPool.commonPool-worker-1
[000012] user= 0.00% sys= 0.00% alloc=     0b/s - ForkJoinPool.commonPool-worker-2
[000013] user= 0.00% sys= 0.01% alloc=     0b/s - Native reference cleanup thread
[000014] user= 0.00% sys= 0.00% alloc=     0b/s - Attach Listener
[000015] user= 0.00% sys= 0.00% alloc=     0b/s - RMI TCP Accept-0
[000017] user= 0.00% sys= 0.00% alloc=     0b/s - RMI Scheduler(0)
```

vs the JDK version that allocates basically nothing:
```
2019-08-03T23:39:41.936-0700 Process summary 
  process cpu=104.07%
  application cpu=100.77% (user=100.42% sys=0.35%)
  other: cpu=3.30% 
  thread count: 9
  heap allocation rate 252kb/s
  safe point rate: 0.8 (events/s) avg. safe point pause: 0.12ms
  safe point sync time: 0.00% processing time: 0.01% (wallclock time)
[000013] user= 0.59% sys= 0.23% alloc=  248kb/s - RMI TCP Connection(1)-127.0.0.1
[000015] user= 0.00% sys= 0.04% alloc=  4257b/s - JMX server connection timeout 15
[000001] user=99.83% sys= 0.08% alloc=     0b/s - main
[000002] user= 0.00% sys= 0.00% alloc=     0b/s - Reference Handler
[000003] user= 0.00% sys= 0.00% alloc=     0b/s - Finalizer
[000004] user= 0.00% sys= 0.00% alloc=     0b/s - Signal Dispatcher
[000010] user= 0.00% sys= 0.00% alloc=     0b/s - Attach Listener
[000012] user= 0.00% sys= 0.00% alloc=     0b/s - RMI TCP Accept-0
[000014] user= 0.00% sys= 0.00% alloc=     0b/s - RMI Scheduler(0)
```

JVM version information:
```
java -version
java version "1.8.0_201"
Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
```

**Comment by SalusaSecondus** [comment]
Thank you for identifying this. We'll get right on this and have it patched in the develop branch shortly.

**Comment by jolynch** [comment]
@SalusaSecondus excellent, let me know when there is a new version to test and I can pick it up easily.

Alternatively let me know if you can't get to it and I can write the patch :-)

**Comment by SalusaSecondus** [comment]
The develop branch has been patched. You should be able to build it locally for testing. Please let me know if you have any difficulties.

**Comment by jolynch** [comment]
@SalusaSecondus Alright I got a chance to test this out and while we're no longer allocating a lot of garbage (nice!) it appears that ACCP is still slower. I have been using the following benchmark as the test suite:
```java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.Provider;
import java.security.Security;
import java.util.concurrent.TimeUnit;

public class ACCPTestLonger {
    private static void updateWithInt(MessageDigest digest, int val)
    {
        digest.update((byte) ((val >>> 24) & 0xFF));
        digest.update((byte) ((val >>> 16) & 0xFF));
        digest.update((byte) ((val >>>  8) & 0xFF));
        digest.update((byte) ((val >>>  0) & 0xFF));
    }

    private static void bench(MessageDigest digest, long intsPerBlock, int blocksToDigest)
    {
        long startTime = System.nanoTime();
        for (int i = 0 ; i < blocksToDigest; i++) {
            for (int j = 0; j < intsPerBlock; j++) {
                updateWithInt(digest, j);
            }
            digest.digest();
        }
        long endTime = System.nanoTime();

        System.out.println(String.format(
                "(%d, %d),", intsPerBlock * 4, TimeUnit.MILLISECONDS.convert(endTime - startTime, TimeUnit.NANOSECONDS)
                ));
        // For verifying the hash is the same
        //System.out.println(String.format("%032X", new BigInteger(1, result)));
    }

    public static void main(String[] args) throws NoSuchAlgorithmException {
        int minExponent = 4;
        int maxExponent = 24;
        if (args.length > 0)
            maxExponent = Math.max(minExponent, Integer.parseInt(args[0]));
        long bytesToTest = 1 << maxExponent;

        String[] algos = new String[] {"MD5", "SHA-1"};

        for (Provider provider: Security.getProviders()) {
            for (String algo : algos) {
                if (provider.getService("MessageDigest", algo) == null)
                    continue;

                System.out.println(String.format("Testing %s - %s", algo, provider.getName()));
                System.out.println(String.format("Benchmarking %d bytes of digest", 1 << maxExponent));
                MessageDigest digest = MessageDigest.getInstance(algo, provider);
                // Warm up first to simulate lots of previous digesting
                for (int k = 0; k < 1 << 16; k++)
                    updateWithInt(digest, k);
                digest.digest();

                for (int i = minExponent; i <= maxExponent; i++) {

                    int intsPerBlock = (1 << i) / 4;
                    int blocksToDigest = (int) (bytesToTest / (intsPerBlock * 4));
                    bench(digest, intsPerBlock, blocksToDigest);
                }
            }
        }

    }
}
```
From the profiles I've been extracting I've identified a few bottlenecks:
1. Copying single bytes arrays appears much slower than just directly writing to the buffer
2. The way that biased locking interacts with the syncronized blocks in the `TemplateHashSpi` update methods can cause a significant reduction in speed when Biased locking doesn't kick in properly.

For example here is the run with biased locking forced off (It is somewhat typical to turn this off in heavily multithreaded code like databases, e.g. Cassandra turns it off by default). Before this test I pinned my laptops cores to `3GHz` with `sudo cpupower frequency-set --min 3GHz --max 3GHz -g performance`
```
$ /usr/lib/jvm/java-8-openjdk-amd64/bin/java -XX:-UseBiasedLocking ACCPTestLonger 24
Testing MD5 - AmazonCorrettoCryptoProvider
Benchmarking 16777216 bytes of digest
(16, 719),
(32, 529),
(64, 497),
(128, 469),
(256, 452),
(512, 443),
(1024, 438),
(2048, 440),
(4096, 438),
(8192, 437),
(16384, 437),
(32768, 437),
(65536, 437),
(131072, 437),
(262144, 436),
(524288, 436),
(1048576, 436),
(2097152, 436),
(4194304, 436),
(8388608, 436),
(16777216, 436),
Testing SHA-1 - AmazonCorrettoCryptoProvider
Benchmarking 16777216 bytes of digest
(16, 731),
(32, 603),
(64, 563),
(128, 534),
(256, 516),
(512, 508),
(1024, 508),
(2048, 504),
(4096, 502),
(8192, 501),
(16384, 507),
(32768, 501),
(65536, 501),
(131072, 501),
(262144, 501),
(524288, 500),
(1048576, 500),
(2097152, 500),
(4194304, 501),
(8388608, 501),
(16777216, 504),
Testing MD5 - SUN
Benchmarking 16777216 bytes of digest
(16, 433),
(32, 307),
(64, 296),
(128, 268),
(256, 253),
(512, 246),
(1024, 243),
(2048, 241),
(4096, 240),
(8192, 239),
(16384, 239),
(32768, 250),
(65536, 246),
(131072, 245),
(262144, 254),
(524288, 270),
(1048576, 255),
(2097152, 257),
(4194304, 256),
(8388608, 242),
(16777216, 247),
Testing SHA-1 - SUN
Benchmarking 16777216 bytes of digest
(16, 580),
(32, 366),
(64, 368),
(128, 341),
(256, 309),
(512, 296),
(1024, 289),
(2048, 279),
(4096, 273),
(8192, 312),
(16384, 285),
(32768, 274),
(65536, 280),
(131072, 283),
(262144, 276),
(524288, 307),
(1048576, 288),
(2097152, 277),
(4194304, 282),
(8388608, 284),
(16777216, 278),
```

Whereas if I set the biased locking delay to zero I get:
```
time /usr/lib/jvm/java-8-openjdk-amd64/bin/java XX:BiasedLockingStartupDelay=0 ACCPTestLonger 24    
Testing MD5 - AmazonCorrettoCryptoProvider
Benchmarking 16777216 bytes of digest
(16, 476),
(32, 315),
(64, 280),
(128, 251),
(256, 231),
(512, 227),
(1024, 221),
(2048, 222),
(4096, 219),
(8192, 219),
(16384, 218),
(32768, 219),
(65536, 218),
(131072, 218),
(262144, 218),
(524288, 218),
(1048576, 218),
(2097152, 218),
(4194304, 218),
(8388608, 218),
(16777216, 218),
Testing SHA-1 - AmazonCorrettoCryptoProvider
Benchmarking 16777216 bytes of digest
(16, 526),
(32, 369),
(64, 330),
(128, 300),
(256, 282),
(512, 275),
(1024, 269),
(2048, 269),
(4096, 268),
(8192, 267),
(16384, 266),
(32768, 266),
(65536, 266),
(131072, 266),
(262144, 266),
(524288, 267),
(1048576, 266),
(2097152, 266),
(4194304, 265),
(8388608, 266),
(16777216, 266),
Testing MD5 - SUN
Benchmarking 16777216 bytes of digest
(16, 450),
(32, 327),
(64, 316),
(128, 289),
(256, 274),
(512, 267),
(1024, 273),
(2048, 261),
(4096, 266),
(8192, 268),
(16384, 260),
(32768, 300),
(65536, 270),
(131072, 294),
(262144, 268),
(524288, 270),
(1048576, 294),
(2097152, 291),
(4194304, 283),
(8388608, 286),
(16777216, 301),
Testing SHA-1 - SUN
Benchmarking 16777216 bytes of digest
(16, 589),
(32, 374),
(64, 367),
(128, 315),
(256, 290),
(512, 287),
(1024, 281),
(2048, 279),
(4096, 276),
(8192, 267),
(16384, 267),
(32768, 268),
(65536, 267),
(131072, 267),
(262144, 267),
(524288, 266),
(1048576, 266),
(2097152, 266),
(4194304, 266),
(8388608, 266),
(16777216, 269),
```
My profiling indicated that we were spending a lot of time in array copy so I tried a simple patch to directly write single bytes into the buffer in [`ab744cee`](https://github.com/jolynch/amazon-corretto-crypto-provider/commit/ab744ceeb5e5e51a451bcca511ca583dd26e10cb) and I did see an improvement in performance, especially when the biased lock kicks in:

```
$ /usr/lib/jvm/java-8-openjdk-amd64/bin/java  -XX:BiasedLockingStartupDelay=0 ACCPTestLonger 24
Testing MD5 - AmazonCorrettoCryptoProvider
Benchmarking 16777216 bytes of digest
(16, 383),
(32, 220),
(64, 175),
(128, 150),
(256, 130),
(512, 125),
(1024, 118),
(2048, 120),
(4096, 119),
(8192, 118),
(16384, 118),
(32768, 118),
(65536, 117),
(131072, 117),
(262144, 117),
(524288, 117),
(1048576, 118),
(2097152, 117),
(4194304, 117),
(8388608, 117),
(16777216, 117),
Testing SHA-1 - AmazonCorrettoCryptoProvider
Benchmarking 16777216 bytes of digest
(16, 365),
(32, 229),
(64, 170),
(128, 139),
(256, 120),
(512, 113),
(1024, 108),
(2048, 111),
(4096, 107),
(8192, 106),
(16384, 106),
(32768, 106),
(65536, 106),
(131072, 106),
(262144, 106),
(524288, 106),
(1048576, 105),
(2097152, 105),
(4194304, 106),
(8388608, 106),
(16777216, 106),
Testing MD5 - SUN
Benchmarking 16777216 bytes of digest
(16, 434),
(32, 306),
(64, 295),
(128, 267),
(256, 251),
(512, 244),
(1024, 240),
(2048, 238),
(4096, 237),
(8192, 237),
(16384, 236),
(32768, 236),
(65536, 236),
(131072, 236),
(262144, 236),
(524288, 236),
(1048576, 236),
(2097152, 236),
(4194304, 236),
(8388608, 236),
(16777216, 236),
Testing SHA-1 - SUN
Benchmarking 16777216 bytes of digest
(16, 558),
(32, 361),
(64, 357),
(128, 307),
(256, 285),
(512, 273),
(1024, 268),
(2048, 266),
(4096, 264),
(8192, 263),
(16384, 267),
(32768, 271),
(65536, 268),
(131072, 270),
(262144, 282),
(524288, 288),
(1048576, 267),
(2097152, 283),
(4194304, 273),
(8388608, 271),
(16777216, 264),
```
I did the analysis with and without my patch with and without biased locking enabled (either `-XX:BiasedLockingStartupDelay=0` or `-XX:-UseBiasedLocking` and gathered the following data:

![image](https://user-images.githubusercontent.com/1420460/63460483-7948a880-c40b-11e9-870f-960256835872.png)

![image](https://user-images.githubusercontent.com/1420460/63460491-7ea5f300-c40b-11e9-80e2-956e48affd70.png)

I'm still profiling more to try to find the bottlenecks but from the data it seems to me that:
1. Directly writing into the buffer similar to [`ab744cee`](https://github.com/jolynch/amazon-corretto-crypto-provider/commit/ab744ceeb5e5e51a451bcca511ca583dd26e10cb) could improve performance
2. If we can remove the synchronized blocks from the TemplateSpi methods that would make `ACCP` strictly faster than `SUN`. To be honest I don't really understand how multpile _java_ threads may be accessing those buffers at the same point. From what I can tell `InputBuffer.buff` is a POJO and I don't see how adding `synchronized` in the callers of `InputBuffer` prevents issues with the native methods. Could we potentially get rid of the locking since MessageDigests are inherently single threaded objects? If not I wonder how we can emulate the locking maybe with atomics in a similar way to how lock biasing works without relying on the JVM to do it...

**Comment by jolynch** [comment]
I added the notebook over in https://github.com/jolynch/performance-analysis/blob/master/notebooks/accp_analysis/accp_sun_hash_singlebyte_analysis.ipynb in case someone wants to re-create the graphs with new data. Just plug in the results from the benchmark script above into the data mapping and it should "just work"

**Comment by SalusaSecondus** [comment]
Thank you very much for these numbers. We'll definitely look into ways to incorporate this feedback and improve performance here.

**Comment by SalusaSecondus** [comment]
Version 1.1.1 released with this fix.

**Comment by jolynch** [comment]
@SalusaSecondus thank you very much for the quick patch! I just rolled `1.1.1` out in our load testing Cassandra clusters and am already seeing significant improvements. It appears we're reducing on CPU time of our digesting functions during quorum reads by up to 50% (so we're going from 20% on CPU time to 10% on CPU time according to flamegraphs). I've also been able to enable AES-GCM without any noticeable increase in CPU load, which is an achievement all in itself.
